Initialise LCB variables with default values

peter-b opened the pull request:
Most LCB variables will be initialised with a default value if they are non-optionally typed with a type for which a default value is defined.

For example, this is now valid:

```
variable tList as List
push "foo" onto tList
```

**Note:** This doesn't include default initialisation of out parameters.

runrevmark committed:
[Bug 14659] Add unit tests for LCB variable default initialisation

peter-b committed:
[Bug 14659] libfoundation: Add default values for all built-in types.

Add a default value for all the "built-in" types that libfoundation
understands (null, boolean, number, array, string, list, proper list,
set and data), and a `MCTypeInfoGetDefault()` accessor function.

peter-b committed:
[Bug 14659] Add release note.

peter-b committed:
[Bug 14659] Test untyped variables are initialised to nothing

If you don't specify a type for an LCB variable, the type is inferred
to be `optional any`.  `optional` types have `nothing` as their
default value.  Therefore, untyped variables should be initialised to
`nothing`.

peter-b commented:
I need to go through all the documentation and remove now-unnecessary initialisation, probably.

However, the code with explicit initialisation will still work fine so I don't think that blocks this PR.

livecodefraser commented:
@livecode-vulcan review okay 5f44412

livecode-vulcan commented:
💙 review by @livecodefraser ok 5f44412e6ebd818bce49b657156a97af0703dc04

peter-b commented:
Needs some further work to avoid redundantly initialising slots used by handler parameters.

livecode-vulcan commented:
😎 test success 5f44412e6ebd818bce49b657156a97af0703dc04
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2503)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2513)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2496)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2514)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2520)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2524)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2536)

peter-b committed:
[Bug 14659] libscript: Initialise frame slots with default values.

When creating a new handler frame, each typed slot will be initialized
with the default value of the type - if the type has one.

peter-b commented:
This now correctly avoids initialising slots three times (once to nothing, once to default value, and once for parameter).  It's still not quite right because it needs to default-initialise module-level slots.

peter-b committed:
[Bug 14659] libscript: Initialise module slots with default values.

Initialise module-level variable slots with default values, when
possible.  The information about module-level slot types is only
attached to variable definitions, so it's not possible to do just a
single pass over the slots because there's no way to find out from a
slot index what type is expected to live in it.  Instead, the module
level default values are implemented by initializing slots in two
passes, first to set default values for all variables that have types
associated, and second to set all remaining slots to `kMCNull`.

livecodefraser commented:
@livecode-vulcan review okay 46e5208

livecodefraser commented:
@livecode-vulcan review okay 27f614b

livecode-vulcan commented:
💙 review by @livecodefraser ok 27f614b5d12edc3d336ff9fa96118947ea581c8b

livecode-vulcan commented:
😎 test success 27f614b5d12edc3d336ff9fa96118947ea581c8b
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2519)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2529)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2514)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2530)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2536)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2540)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2552)
